### PR TITLE
59 keep track of agents responsible for edges

### PR DESCRIPTION
**Pull request by cheerstopaula**
Yankee swap version (general_yankee_swap_E) keeping track of agents responsible for edges in order to reduce looping and overall calls to valuation function.

### CODE REVIEW

File: src/fair/allocation.py [hunk 1]
```diff
@@ -24,39 +31,71 @@ def initialize_allocation_matrix(items, agents):
...
+    """Initialize array that keeps track of the agents responsible for a specific edge between two
+    different items on the exchange graph.
+    Initially, there are no edges, thus edge matrix is a len(items) x len(items) array of empty lists.
+
+    Args:
+       N (int): Number of items
+
+    Returns:
+        list: N x N list of empty lists
+    """
```
Comment (gbiss) [inline src/fair/allocation.py:44]: I think you could do this with a one-liner:

```
[[[] for i in range(N)] for j in range(N)]
```

File: src/fair/allocation.py [hunk 2]
```diff
@@ -24,39 +31,71 @@ def initialize_allocation_matrix(items, agents):
     return X
 
 
-def initialize_exchange_graph(items):
+def initialize_edge_matrix(N: int):
+    """Initialize array that keeps track of the agents responsible for a specific edge between two
+    different items on the exchange graph.
+    Initially, there are no edges, thus edge matrix is a len(items) x len(items) array of empty lists.
+
```
Comment (gbiss) [inline src/fair/allocation.py:38]: For docstring, please try to stick to:

```
"""Short description

A longer description after, which could span
several lines.
"""
```

File: src/fair/allocation.py [hunk 3]
```diff
@@ -74,16 +113,33 @@ def get_gain_function(X, agents, items, agent_picked, criteria, weights):
         return w_i / (val + 1)
 
 
-"""Retrieve information or update current allocation"""
+def get_owners_list(X: type[np.ndarray], item_index: int):
+    """From the exchange matrix, list of indeces of all agents that currently have certain item.
```
Comment (gbiss) [inline src/fair/allocation.py:117]: "indeces" -> "indices"

File: src/fair/allocation.py [hunk 4]
```diff
@@ -24,39 +31,71 @@ def initialize_allocation_matrix(items, agents):
...
+    X: type[np.ndarray],
+    agents: list[BaseAgent],
+    items: list[ScheduleItem],
+    agent_picked: int,
+    criteria: str,
+    weights: list[float],
+):
+    """Get gain function for a certain agent from the current allocation, according to the gain function criteria and agent assigned weight.
+    This function is used to update the gain function for the item that just played, to keep track of priority agents.
+
```
Comment (gbiss) [inline src/fair/allocation.py:87]: "item that just played" or "agent that just played"?

File: src/fair/allocation.py [hunk 5]
```diff
@@ -74,16 +113,33 @@ def get_gain_function(X, agents, items, agent_picked, criteria, weights):
...
+    """From the exchange matrix, list of indeces of all agents that currently have certain item.
 
+    Args:
+        X (type[np.ndarray]): Allocation matrix
+        item_index (int): index of the item for which we want to get the owners
 
-def get_owners_list(X, item_index):
+    Returns:
+        list[int]: list of item's owners' indeces
+    """
```
Comment (gbiss) [inline src/fair/allocation.py:125]: "indices"^

File: src/fair/allocation.py [hunk 6]
```diff
@@ -249,13 +448,81 @@ def update_exchange_graph(X, G, path_og, agents, items, agents_involved):
...
+    X: type[np.ndarray],
+    G: type[nx.Graph],
+    E: list[list],
+    agents: list[BaseAgent],
+    items: list[ScheduleItem],
+    path_og: list[int],
+    agents_involved: list[int],
+):
+    """_summary_
+
```
Comment (gbiss) [inline src/fair/allocation.py:461]: Please fill in summary^

File: src/fair/allocation.py [hunk 7]
```diff
@@ -424,24 +677,44 @@ def original_yankee_swap(agents, items, plot_exchange_graph=False):
...
+    items: list[ScheduleItem],
+    criteria: str = "LorenzDominance",
+    weights: list = [],
+    plot_exchange_graph: bool = False,
 ):
-    """General Yankee Swap"""
-    players = initialize_players(agents)
+    """General Yankee swap allocation algorithm, edge matrix version
+    Equivalent to general_yankee_swap, just different book keeping to speed things up
+
```
Comment (gbiss) [inline src/fair/allocation.py:689]: "bookkeeping" one word

File: src/fair/allocation.py (hunk 1 above)
Comment (cheerstopaula) [reply src/fair/allocation.py:44]: Changed (eliminated the function altogether)

File: src/fair/allocation.py (hunk 4 above)
Comment (cheerstopaula) [reply src/fair/allocation.py:87]: yesss good catch. Changed!

**Review by gbiss — APPROVED** [review]

**Pull request merged**